codec/enforceproto/extraneous_fields.go

```diff
@@ -0,0 +1,356 @@
+	}
+
+	// Now cache it.
+	protoMessageForTypeNameMu.Lock()
+	protoMessageForTypeNameCache[protoMessageName] = msg
+	protoMessageForTypeNameMu.Unlock()
+
+	return msg, nil
+}
+
+// checks is a mapping of protowire.Type to supported descriptor.FieldDescriptorProto_Type.
+// it is implemented this way so as to have constant time lookups and avoid the overhead
+// from O(n) walking of switch. The change to using this mapping boosts throughput by about 200%.
+var checks = [...]map[descriptor.FieldDescriptorProto_Type]bool{
+	// "0	Varint: int32, int64, uint32, uint64, sint32, sint64, bool, enum"
+	0: {
+		descriptor.FieldDescriptorProto_TYPE_INT32:  true,
+		descriptor.FieldDescriptorProto_TYPE_INT64:  true,
+		descriptor.FieldDescriptorProto_TYPE_UINT32: true,
+		descriptor.FieldDescriptorProto_TYPE_UINT64: true,
+		descriptor.FieldDescriptorProto_TYPE_SINT32: true,
+		descriptor.FieldDescriptorProto_TYPE_SINT64: true,
+		descriptor.FieldDescriptorProto_TYPE_BOOL:   true,
+		descriptor.FieldDescriptorProto_TYPE_ENUM:   true,
+	},
+
+	// "1	64-bit:	fixed64, sfixed64, double"
+	1: {
+		descriptor.FieldDescriptorProto_TYPE_FIXED64:  true,
+		descriptor.FieldDescriptorProto_TYPE_SFIXED64: true,
+		descriptor.FieldDescriptorProto_TYPE_DOUBLE:   true,
+	},
+
+	// "2	Length-delimited: string, bytes, embedded messages, packed repeated fields"
+	2: {
+		descriptor.FieldDescriptorProto_TYPE_STRING:  true,
+		descriptor.FieldDescriptorProto_TYPE_BYTES:   true,
+		descriptor.FieldDescriptorProto_TYPE_MESSAGE: true,
+	},
+
+	// "3	Start group:	groups (deprecated)"
+	3: {
+		descriptor.FieldDescriptorProto_TYPE_GROUP: true,
+	},
+
+	// "4	End group:	groups (deprecated)"
+	4: {
+		descriptor.FieldDescriptorProto_TYPE_GROUP: true,
+	},
+
+	// "5	32-bit:	fixed32, sfixed32, float"
+	5: {
+		descriptor.FieldDescriptorProto_TYPE_FIXED32:  true,
+		descriptor.FieldDescriptorProto_TYPE_SFIXED32: true,
+		descriptor.FieldDescriptorProto_TYPE_FLOAT:    true,
+	},
+}
+
+// canEncodeType returns true if the wireType is suitable for encoding the descriptor type.
+// See https://developers.google.com/protocol-buffers/docs/encoding#structure.
+func canEncodeType(wireType protowire.Type, descType descriptor.FieldDescriptorProto_Type) bool {
+	if iwt := int(wireType); iwt < 0 || iwt >= len(checks) {
+		return false
+	}
+	return checks[wireType][descType]
+}
+
+// ErrMismatchedWireType describes a mismatch between
+// expected and got wireTypes for a specific tag number.
+type ErrMismatchedWireType struct {
+	Type         string
+	GotWireType  protowire.Type
+	WantWireType protowire.Type
+	TagNum       protowire.Number
+}
+
+func (mwt *ErrMismatchedWireType) String() string {
+	return fmt.Sprintf("Mismatched %q: {TagNum: %d, GotWireType: %q != WantWireType: %q}",
+		mwt.Type, mwt.TagNum, wireTypeToString(mwt.GotWireType), wireTypeToString(mwt.WantWireType))
+}
+
+func (mwt *ErrMismatchedWireType) Error() string {
```

Comment:
```suggestion
// String implements the error interface
func (mwt *ErrMismatchedWireType) String() string {
	return fmt.Sprintf("Mismatched %q: {TagNum: %d, GotWireType: %q != WantWireType: %q}",
		mwt.Type, mwt.TagNum, wireTypeToString(mwt.GotWireType), wireTypeToString(mwt.WantWireType))
}

// Error implements the error interface
func (mwt *ErrMismatchedWireType) Error() string {
```

Reply:
Not really though, we aren't implementing the error interface :) The Error() method just happens to use .String()